src/main/java/dev/openfeature/sdk/ProviderRepository.java

```diff
@@ -0,0 +1,156 @@
+        }
+        initializeProvider(provider);
+    }
+
+    /**
+     * Add a provider for a named client.
+     *
+     * @param clientName The name of the client.
+     * @param provider   The provider to set.
+     */
+    public void setProvider(String clientName, FeatureProvider provider) {
+        if (provider == null) {
+            throw new IllegalArgumentException("Provider cannot be null");
+        }
+        if (clientName == null) {
+            throw new IllegalArgumentException("clientName cannot be null");
+        }
+        initializeProvider(clientName, provider);
+    }
+
+    private void initializeProvider(FeatureProvider provider) {
+        initializingDefaultProvider = provider;
+        initializeProvider(provider, this::updateDefaultProviderAfterInitialization);
+    }
+
+    private void initializeProvider(String clientName, FeatureProvider provider) {
+        initializingNamedProviders.put(clientName, provider);
+        initializeProvider(provider, newProvider -> updateProviderAfterInit(clientName, newProvider));
+    }
+
+    private void initializeProvider(FeatureProvider provider, Consumer<FeatureProvider> afterInitialization) {
+        taskExecutor.submit(() -> {
+            try {
+                if (isProviderNotBoundMultipleTimes(provider)) {
+                    provider.initialize();
+                }
+                afterInitialization.accept(provider);
+            } catch (Exception e) {
+                log.error("Exception when initializing feature provider {}", provider.getClass().getName(), e);
+            }
+        });
+    }
+
+    private void updateProviderAfterInit(String clientName, FeatureProvider newProvider) {
+        Optional
+                .ofNullable(initializingNamedProviders.get(clientName))
+                .filter(initializingProvider -> initializingProvider.equals(newProvider))
+                .ifPresent(provider -> updateNamedProviderAfterInitialization(clientName, provider));
+    }
+
+    private void updateDefaultProviderAfterInitialization(FeatureProvider initializedProvider) {
+        Optional
+                .ofNullable(this.initializingDefaultProvider)
+                .filter(initializingProvider -> initializingProvider.equals(initializedProvider))
+                .ifPresent(this::replaceDefaultProvider);
+    }
+
+    private void replaceDefaultProvider(FeatureProvider provider) {
+        FeatureProvider oldProvider = this.defaultProvider.getAndSet(provider);
+        if (isOldProviderNotBoundByName(oldProvider)) {
+            shutdownProvider(oldProvider);
+        }
+    }
+
+    private boolean isOldProviderNotBoundByName(FeatureProvider oldProvider) {
+        return !this.providers.containsValue(oldProvider);
+    }
+
+    private void updateNamedProviderAfterInitialization(String clientName, FeatureProvider initializedProvider) {
+        Optional
+                .ofNullable(this.initializingNamedProviders.get(clientName))
```

Comment:
```suggestion
                .ofNullable(this.initializingNamedProviders.remove(clientName))
```

This allows us to remove the dedicated remove call in `replaceNamedProviderAndShutdownOldOne` [1]. And this reduce complexity of the code and avoid future mistakes 


[1] - https://github.com/open-feature/java-sdk/blob/a6506558e0ab08ce2db9f7cae6c3440bab58066f/src/main/java/dev/openfeature/sdk/ProviderRepository.java#L123

Reply:
we can't do this, as a new provider might be bound to the client name already, see the `filter` method in line 117. that's also the reason i chose `Map.remove(key, value)` and not simply `Map.remove(key)` in line 123.